// @yellicode/csharp/dist/es6/comment-writer.js
import { CodeWriterUtility } from '@yellicode/core';
export class CSharpCommentWriter {
    constructor(writer, maxCommentWidth) {
        this.writer = writer;
        this.maxCommentWidth = maxCommentWidth;
    }
    writeDelimitedCommentParagraph(text) {
        if (text == null)
            return;
        const lines = [text];
        this.writer.writeLine("/*");
        this.writeCommentLines(lines, "* ");
        this.writer.writeLine("*/");
    }
    writeDelimitedCommentLines(lines) {
        if (lines == null)
            return;
        this.writer.writeLine("/*");
        this.writeCommentLines(lines, "* ");
        this.writer.writeLine("*/");
    }
    writeCommentLines(lines, prefix) {
        if (!lines)
            return;
        lines.forEach(line => {
            if (line == null)
                return;
            if (this.maxCommentWidth > 0 && line.length > this.maxCommentWidth) {
                // See if we can split the line
                const split = CodeWriterUtility.wordWrap(line, this.maxCommentWidth);
                split.forEach(s => {
                    this.writer.writeLine(`${prefix}${s}`);
                });
            }
            else
                this.writer.writeLine(`${prefix}${line}`);
        });
    }
}
