// @yellicode/csharp/dist/es6/definition-builder.js
import * as elements from '@yellicode/elements';
import * as opts from './options';
import { ParameterDirectionKind } from '@yellicode/elements';
import { CSharpTypeNameProvider } from './csharp-type-name-provider';
export class DefinitionBuilder {
    constructor(typeNameProvider) {
        this.typeNameProvider = typeNameProvider;
    }
    buildNamespaceDefinition(pack, options) {
        if (!options)
            options = {};
        // We have no namespace features yet, but the NamespaceFeatures type is there for future extension
        const features = (options.features === undefined) ? opts.NamespaceFeatures.All : options.features;
        let name;
        if (options.writeFullName) {
            const allPackages = pack.getNestingPackages();
            allPackages.push(pack); // add the package itself
            name = allPackages.map(p => p.name).join('.');
        }
        else
            name = pack.name;
        return { name: name };
    }
    buildClassDefinition(type, options) {
        // Initialize options and features
        if (!options)
            options = {};
        const features = (options.features === undefined) ? opts.ClassFeatures.All : options.features;
        // Build the base definition
        const definition = DefinitionBuilder.buildTypeDefinition(type, !!(features & opts.ClassFeatures.XmlDocSummary), options);
        // Build the class-specific definition
        definition.inherits = DefinitionBuilder.buildInherits(type, options.inherits);
        definition.implements = DefinitionBuilder.buildImplements(type, options.implements);
        if (elements.isClass(type)) {
            definition.isAbstract = type.isAbstract || options.isAbstract;
        }
        return definition;
    }
    buildStructDefinition(type, options) {
        // Initialize options and features
        if (!options)
            options = {};
        const features = (options.features === undefined) ? opts.StructFeatures.All : options.features;
        // Build the base definition
        const definition = DefinitionBuilder.buildTypeDefinition(type, !!(features & opts.StructFeatures.XmlDocSummary), options);
        // Build the struct-specific definition        
        definition.implements = DefinitionBuilder.buildImplements(type, options.implements);
        return definition;
    }
    buildInterfaceDefinition(type, options) {
        // Initialize options and features
        if (!options)
            options = {};
        const features = (options.features === undefined) ? opts.InterfaceFeatures.All : options.features;
        // Build the base definition
        const definition = DefinitionBuilder.buildTypeDefinition(type, !!(features & opts.InterfaceFeatures.XmlDocSummary), options);
        // Build the interface-specific definition
        definition.inherits = DefinitionBuilder.buildInherits(type, options.inherits);
        return definition;
    }
    buildEnumMemberDefinition(literal, isLast, options) {
        // Initialize options and features
        if (!options)
            options = {};
        const features = (options.features === undefined) ? opts.EnumMemberFeatures.All : options.features;
        const buildInitializers = !!(features & opts.EnumFeatures.Initializers);
        // Build the base definition
        const definition = DefinitionBuilder.buildDefinitionBase(literal, !!(features & opts.EnumMemberFeatures.XmlDocSummary));
        // Build the member-specific definition
        definition.isLast = isLast || false;
        if (buildInitializers && literal.specification != null) { // using '!=' on purpose  
            const specification = literal.specification;
            definition.value = elements.isLiteralInteger(specification) ? specification.value : undefined;
        }
        return definition;
    }
    buildEnumDefinition(type, options) {
        // Initialize options and features
        if (!options)
            options = {};
        const features = (options.features === undefined) ? opts.EnumFeatures.All : options.features;
        // Build the base definition         
        const definition = DefinitionBuilder.buildDefinitionBase(type, !!(features & opts.EnumFeatures.XmlDocSummary));
        // Build the enum-specific definition
        definition.accessModifier = DefinitionBuilder.getAccessModifierString(type.visibility);
        // Build enum members
        if (elements.isEnumeration(type) && type.ownedLiterals) {
            const members = [];
            // Pass on enum features to enum member features
            let memberFeatures = opts.EnumMemberFeatures.None;
            if (features & opts.EnumFeatures.XmlDocSummary)
                memberFeatures |= opts.EnumMemberFeatures.XmlDocSummary;
            if (features & opts.EnumFeatures.Initializers)
                memberFeatures |= opts.EnumMemberFeatures.Initializers;
            type.ownedLiterals.forEach((literal, index) => {
                let isLast = index === type.ownedLiterals.length - 1;
                const member = this.buildEnumMemberDefinition(literal, isLast, { features: memberFeatures });
                members.push(member);
            });
            definition.members = members;
        }
        return definition;
    }
    buildPropertyDefinition(property, options) {
        // Initialize options and features
        if (!options)
            options = {};
        const features = (options.features === undefined) ? opts.PropertyFeatures.All : options.features;
        const ownerIsInterface = elements.isInterface(property.owner);
        // Build the base definition         
        const definition = DefinitionBuilder.buildDefinitionBase(property, !!(features & opts.PropertyFeatures.XmlDocSummary));
        // Build the property-specific definition
        const typename = this.getFullTypeName(property, options.collectionType || null, 'object');
        definition.accessModifier = DefinitionBuilder.getAccessModifierString(property.visibility);
        definition.isVirtual = options.virtual;
        definition.typeName = typename;
        if (!ownerIsInterface && (features & opts.PropertyFeatures.AccessModifier))
            definition.accessModifier = DefinitionBuilder.getAccessModifierString(property.visibility);
        if ((features & opts.PropertyFeatures.OptionalModifier) &&
            property.lower === 0 &&
            CSharpTypeNameProvider.canBeNullable(property, typename)) {
            definition.isNullable = true;
        }
        definition.noSetter = property.isReadOnly || property.isDerived;
        definition.defaultValue = DefinitionBuilder.getDefaultValueString(property.defaultValue);
        return definition;
    }
    buildMethodDefinition(operation, options) {
        if (!options)
            options = {};
        const features = (options.features === undefined) ? opts.MethodFeatures.All : options.features;
        const ownerIsInterface = elements.isInterface(operation.owner);
        // Build the base definition         
        const definition = DefinitionBuilder.buildDefinitionBase(operation, !!(features & opts.MethodFeatures.XmlDocSummary));
        definition.isConstructor = operation.isConstructor;
        definition.isStatic = operation.isStatic;
        definition.isPartial = options.isPartial;
        if (!operation.isConstructor) {
            definition.isAbstract = options.isAbstract || (!options.isVirtual && operation.isAbstract);
            definition.isVirtual = options.isVirtual;
        }
        if (!ownerIsInterface)
            definition.accessModifier = DefinitionBuilder.getAccessModifierString(operation.visibility);
        // Get the return type and documentation
        if (!operation.isConstructor) {
            var returnParameter = operation.getReturnParameter();
            if (returnParameter) {
                definition.returnTypeName = this.getFullTypeName(returnParameter, options.collectionType || null);
                if (features & opts.MethodFeatures.XmlDocReturns) {
                    definition.xmlDocReturns = DefinitionBuilder.buildXmlDocSummary(operation);
                }
            }
        }
        // Build parameter definitions      
        definition.parameters = this.buildParameterDefinitions(operation.ownedParameters, options);
        return definition;
    }
    buildParameterDefinitions(params, options) {
        if (!options)
            options = {};
        const features = (options.features === undefined) ? opts.MethodFeatures.All : options.features;
        const inOutParameters = [];
        if (!params)
            return inOutParameters;
        params.forEach(p => {
            if (p.direction === ParameterDirectionKind.return)
                return;
            const typeName = this.getFullTypeName(p, options.collectionType || null, 'object');
            const paramDefinition = DefinitionBuilder.buildDefinitionBase(p, !!(features & opts.MethodFeatures.XmlDocParameters));
            paramDefinition.isOutput = p.direction === elements.ParameterDirectionKind.out;
            paramDefinition.isReference = p.direction === elements.ParameterDirectionKind.inout;
            paramDefinition.isNullable = p.lower === 0 && CSharpTypeNameProvider.canBeNullable(p, typeName);
            paramDefinition.typeName = typeName;
            paramDefinition.defaultValue = DefinitionBuilder.getDefaultValueString(p.defaultValue),
                inOutParameters.push(paramDefinition);
        });
        return inOutParameters;
    }
    static buildInherits(type, additional) {
        if (!elements.isClassifier(type)) {
            return;
        }
        const allNames = [];
        if (type.generalizations) {
            // todo: allow qualifiedName
            allNames.push(...type.generalizations.map(g => g.general.name));
        }
        if (additional) {
            allNames.push(...additional);
        }
        return allNames.length ? allNames : undefined;
    }
    static buildImplements(type, additional) {
        if (!elements.isBehavioredClassifier(type)) {
            return;
        }
        const allNames = [];
        if (type.interfaceRealizations) {
            // todo: allow qualifiedName 
            allNames.push(...type.interfaceRealizations.map(ir => ir.contract.name));
        }
        if (additional) {
            allNames.push(...additional);
        }
        return allNames.length ? allNames : undefined;
    }
    static buildXmlDocSummary(element) {
        if (!element.ownedComments || !element.ownedComments.length) {
            return undefined;
        }
        return element.ownedComments.map(c => c.body);
    }
    static buildTypeDefinition(type, buildXmlDocSummary, options) {
        var definition = DefinitionBuilder.buildDefinitionBase(type, buildXmlDocSummary);
        definition.accessModifier = DefinitionBuilder.getAccessModifierString(type.visibility);
        definition.isPartial = options.isPartial;
        return definition;
    }
    static buildDefinitionBase(element, buildXmlDocSummary) {
        var definition = {
            name: element.name,
            xmlDocSummary: buildXmlDocSummary ? DefinitionBuilder.buildXmlDocSummary(element) : undefined
        };
        return definition;
    }
    static getDefaultValueString(defaultValue) {
        if (!defaultValue)
            return undefined;
        return elements.isLiteralString(defaultValue) ?
            `"${defaultValue.value}"` :
            defaultValue.getStringValue();
    }
    static getAccessModifierString(visibility) {
        switch (visibility) {
            case elements.VisibilityKind.public:
                return 'public';
            case elements.VisibilityKind.private:
                return 'private';
            case elements.VisibilityKind.protected:
                return 'protected';
            case elements.VisibilityKind.package:
                return 'internal';
            default:
                return undefined;
        }
    }
    getFullTypeName(typedElement, collectionType, fallback) {
        const typeName = this.typeNameProvider.getTypeName(typedElement) || fallback;
        if (!typeName)
            return; // no type name and no fallback
        if (elements.isMultiplicityElement(typedElement) && typedElement.isMultivalued()) {
            switch (collectionType) {
                case opts.CollectionType.IList:
                    return `IList<${typeName}>`;
                case opts.CollectionType.IEnumerable:
                    return `IEnumerable<${typeName}>`;
                default:
                    return `ICollection<${typeName}>`;
            }
        }
        else
            return typeName;
    }
}
