// @yellicode/csharp/dist/es6/model-builder.js
import * as elements from '@yellicode/elements';
import * as opts from './options';
export class ModelBuilder {
    static buildClassDefinition(type, options) {
        if (!options)
            options = {};
        const features = (options.features === undefined) ? opts.ClassFeatures.All : options.features;
        const definition = { name: type.name };
        if (features & opts.ClassFeatures.XmlDocSummary) {
            definition.xmlDocSummary = ModelBuilder.buildXmlDocSummary(type);
        }
        definition.isPartial = options.isPartial;
        if (elements.isClassifier(type)) {
            definition.inherits = ModelBuilder.buildInherits(type, options.inherits);
        }
        if (elements.isClass(type)) {
            definition.isAbstract = type.isAbstract;
            definition.implements = ModelBuilder.buildImplements(type, options.implements);
        }
        return definition;
    }
    static buildInterfaceDefinition(type, options) {
        if (!options)
            options = {};
        const features = (options.features === undefined) ? opts.InterfaceFeatures.All : options.features;
        const definition = { name: type.name };
        if (features & opts.InterfaceFeatures.XmlDocSummary) {
            definition.xmlDocSummary = ModelBuilder.buildXmlDocSummary(type);
        }
        definition.isPartial = options.isPartial;
        if (elements.isClassifier(type)) {
            definition.inherits = ModelBuilder.buildInherits(type, options.inherits);
        }
        return definition;
    }
    static buildInherits(type, additional) {
        const allNames = [];
        if (type.generalizations) {
            // todo: allow qualifiedName
            allNames.push(...type.generalizations.map(g => g.general.name));
        }
        if (additional) {
            allNames.push(...additional);
        }
        return allNames.length ? allNames : undefined;
    }
    static buildImplements(type, additional) {
        const allNames = [];
        if (type.interfaceRealizations) {
            // todo: allow qualifiedName 
            allNames.push(...type.interfaceRealizations.map(ir => ir.contract.name));
        }
        if (additional) {
            allNames.push(...additional);
        }
        return allNames.length ? allNames : undefined;
    }
    static buildXmlDocSummary(element) {
        if (!element.ownedComments || !element.ownedComments.length) {
            return;
        }
        return element.ownedComments.map(c => c.body);
    }
}
