// @yellicode/csharp/dist/es6/options.js
/**
 * Enumerates the possible collection types to generated for properties and parameters.
 */
export var CollectionType;
(function (CollectionType) {
    CollectionType[CollectionType["ICollection"] = 0] = "ICollection";
    CollectionType[CollectionType["IEnumerable"] = 1] = "IEnumerable";
    CollectionType[CollectionType["IList"] = 2] = "IList";
})(CollectionType || (CollectionType = {}));
export var NamespaceFeatures;
(function (NamespaceFeatures) {
    NamespaceFeatures[NamespaceFeatures["None"] = 0] = "None";
    NamespaceFeatures[NamespaceFeatures["All"] = 0] = "All";
})(NamespaceFeatures || (NamespaceFeatures = {}));
export var ClassFeatures;
(function (ClassFeatures) {
    ClassFeatures[ClassFeatures["None"] = 0] = "None";
    ClassFeatures[ClassFeatures["XmlDocSummary"] = 1] = "XmlDocSummary";
    ClassFeatures[ClassFeatures["Generalizations"] = 2] = "Generalizations";
    ClassFeatures[ClassFeatures["InterfaceRealizations"] = 4] = "InterfaceRealizations";
    ClassFeatures[ClassFeatures["All"] = 7] = "All";
})(ClassFeatures || (ClassFeatures = {}));
export var StructFeatures;
(function (StructFeatures) {
    StructFeatures[StructFeatures["None"] = 0] = "None";
    StructFeatures[StructFeatures["XmlDocSummary"] = 1] = "XmlDocSummary";
    StructFeatures[StructFeatures["InterfaceRealizations"] = 2] = "InterfaceRealizations";
    StructFeatures[StructFeatures["All"] = 3] = "All";
})(StructFeatures || (StructFeatures = {}));
export var InterfaceFeatures;
(function (InterfaceFeatures) {
    InterfaceFeatures[InterfaceFeatures["None"] = 0] = "None";
    InterfaceFeatures[InterfaceFeatures["XmlDocSummary"] = 1] = "XmlDocSummary";
    InterfaceFeatures[InterfaceFeatures["Generalizations"] = 2] = "Generalizations";
    InterfaceFeatures[InterfaceFeatures["All"] = 3] = "All";
})(InterfaceFeatures || (InterfaceFeatures = {}));
export var EnumFeatures;
(function (EnumFeatures) {
    EnumFeatures[EnumFeatures["None"] = 0] = "None";
    EnumFeatures[EnumFeatures["XmlDocSummary"] = 1] = "XmlDocSummary";
    /**
     * Writes enumeration member initializers. This flag only applies when values are provided in the model.
     */
    EnumFeatures[EnumFeatures["Initializers"] = 2] = "Initializers";
    EnumFeatures[EnumFeatures["All"] = 3] = "All";
})(EnumFeatures || (EnumFeatures = {}));
export var EnumMemberFeatures;
(function (EnumMemberFeatures) {
    EnumMemberFeatures[EnumMemberFeatures["None"] = 0] = "None";
    EnumMemberFeatures[EnumMemberFeatures["XmlDocSummary"] = 1] = "XmlDocSummary";
    /**
     * Writes enumeration member initializers. This flag only applies when values are provided in the model.
     */
    EnumMemberFeatures[EnumMemberFeatures["Initializers"] = 2] = "Initializers";
    EnumMemberFeatures[EnumMemberFeatures["All"] = 3] = "All";
})(EnumMemberFeatures || (EnumMemberFeatures = {}));
export var PropertyFeatures;
(function (PropertyFeatures) {
    PropertyFeatures[PropertyFeatures["None"] = 0] = "None";
    PropertyFeatures[PropertyFeatures["XmlDocSummary"] = 1] = "XmlDocSummary";
    /**
     * The access modifier if the owner is not an Interface.
     */
    PropertyFeatures[PropertyFeatures["AccessModifier"] = 2] = "AccessModifier";
    PropertyFeatures[PropertyFeatures["OptionalModifier"] = 4] = "OptionalModifier";
    PropertyFeatures[PropertyFeatures["All"] = 7] = "All";
})(PropertyFeatures || (PropertyFeatures = {}));
export var MethodFeatures;
(function (MethodFeatures) {
    MethodFeatures[MethodFeatures["None"] = 0] = "None";
    MethodFeatures[MethodFeatures["XmlDocSummary"] = 1] = "XmlDocSummary";
    MethodFeatures[MethodFeatures["XmlDocParameters"] = 2] = "XmlDocParameters";
    MethodFeatures[MethodFeatures["XmlDocReturns"] = 4] = "XmlDocReturns";
    /**
    * The access modifier if the owner is not an Interface.
    */
    MethodFeatures[MethodFeatures["AccessModifier"] = 8] = "AccessModifier";
    MethodFeatures[MethodFeatures["All"] = 15] = "All";
})(MethodFeatures || (MethodFeatures = {}));
