// @yellicode/typescript/dist/es5/options.js
/**
 * Defines options for dealing with elements that are 'optional', meaning that they have a lower bound of 0.
 */
export var OptionalityModifier;
(function (OptionalityModifier) {
    /**
     * Don't write any modifier.
     */
    OptionalityModifier[OptionalityModifier["Ignore"] = 0] = "Ignore";
    /**
     * Write a 'null' keyword, for example 'myProperty: string | null;';
     */
    OptionalityModifier[OptionalityModifier["NullKeyword"] = 1] = "NullKeyword";
    /**
     * Write a question token, for example 'myProperty?: string;';
     */
    OptionalityModifier[OptionalityModifier["QuestionToken"] = 2] = "QuestionToken";
})(OptionalityModifier || (OptionalityModifier = {}));
export var EnumFeatures;
(function (EnumFeatures) {
    EnumFeatures[EnumFeatures["None"] = 0] = "None";
    EnumFeatures[EnumFeatures["JsDocDescription"] = 1] = "JsDocDescription";
    EnumFeatures[EnumFeatures["Initializers"] = 2] = "Initializers";
    EnumFeatures[EnumFeatures["All"] = 3] = "All";
})(EnumFeatures || (EnumFeatures = {}));
export var ClassFeatures;
(function (ClassFeatures) {
    ClassFeatures[ClassFeatures["None"] = 0] = "None";
    ClassFeatures[ClassFeatures["JsDocDescription"] = 1] = "JsDocDescription";
    ClassFeatures[ClassFeatures["Generalizations"] = 2] = "Generalizations";
    ClassFeatures[ClassFeatures["InterfaceRealizations"] = 4] = "InterfaceRealizations";
    ClassFeatures[ClassFeatures["All"] = 7] = "All";
})(ClassFeatures || (ClassFeatures = {}));
export var InterfaceFeatures;
(function (InterfaceFeatures) {
    InterfaceFeatures[InterfaceFeatures["None"] = 0] = "None";
    InterfaceFeatures[InterfaceFeatures["JsDocDescription"] = 1] = "JsDocDescription";
    InterfaceFeatures[InterfaceFeatures["Generalizations"] = 2] = "Generalizations";
    InterfaceFeatures[InterfaceFeatures["All"] = 3] = "All";
})(InterfaceFeatures || (InterfaceFeatures = {}));
export var PropertyFeatures;
(function (PropertyFeatures) {
    PropertyFeatures[PropertyFeatures["None"] = 0] = "None";
    PropertyFeatures[PropertyFeatures["JsDocDescription"] = 1] = "JsDocDescription";
    PropertyFeatures[PropertyFeatures["AccessModifier"] = 2] = "AccessModifier";
    PropertyFeatures[PropertyFeatures["ReadonlyModifier"] = 4] = "ReadonlyModifier";
    PropertyFeatures[PropertyFeatures["OptionalModifier"] = 8] = "OptionalModifier";
    PropertyFeatures[PropertyFeatures["Initializer"] = 16] = "Initializer";
    /**
     * Writes the so-called 'definite assignment assertion modifier' for the property if the property is required.
     * There are certain scenarios where properties can be initialized indirectly (perhaps by a helper method or dependency injection library),
     * in which case you can use the definite assignment assertion modifiers for your properties.
     */
    PropertyFeatures[PropertyFeatures["DefiniteAssignmentAssertionModifier"] = 32] = "DefiniteAssignmentAssertionModifier";
    PropertyFeatures[PropertyFeatures["All"] = 63] = "All";
    /**
     * Include all features, except the ones that are not usable on interface properties.
     */
    PropertyFeatures[PropertyFeatures["AllInterfaceProperty"] = 13] = "AllInterfaceProperty";
})(PropertyFeatures || (PropertyFeatures = {}));
export var FunctionFeatures;
(function (FunctionFeatures) {
    FunctionFeatures[FunctionFeatures["None"] = 0] = "None";
    FunctionFeatures[FunctionFeatures["JsDocDescription"] = 1] = "JsDocDescription";
    FunctionFeatures[FunctionFeatures["AccessModifier"] = 2] = "AccessModifier";
    FunctionFeatures[FunctionFeatures["OptionalModifier"] = 4] = "OptionalModifier";
    FunctionFeatures[FunctionFeatures["All"] = 7] = "All";
})(FunctionFeatures || (FunctionFeatures = {}));
export var ParameterFeatures;
(function (ParameterFeatures) {
    ParameterFeatures[ParameterFeatures["None"] = 0] = "None";
    ParameterFeatures[ParameterFeatures["JsDocDescription"] = 1] = "JsDocDescription";
    ParameterFeatures[ParameterFeatures["OptionalModifier"] = 2] = "OptionalModifier";
    ParameterFeatures[ParameterFeatures["All"] = 3] = "All";
})(ParameterFeatures || (ParameterFeatures = {}));
