// @yellicode/csharp/dist/es5/comment-writer.js
import { CodeWriterUtility } from '@yellicode/core';
var CSharpCommentWriter = /** @class */ (function () {
    function CSharpCommentWriter(writer, maxCommentWidth) {
        this.writer = writer;
        this.maxCommentWidth = maxCommentWidth;
    }
    CSharpCommentWriter.prototype.writeDelimitedCommentParagraph = function (text) {
        if (text == null)
            return;
        var lines = [text];
        this.writer.writeLine("/*");
        this.writeCommentLines(lines, "* ");
        this.writer.writeLine("*/");
    };
    CSharpCommentWriter.prototype.writeDelimitedCommentLines = function (lines) {
        if (lines == null)
            return;
        this.writer.writeLine("/*");
        this.writeCommentLines(lines, "* ");
        this.writer.writeLine("*/");
    };
    CSharpCommentWriter.prototype.writeCommentLines = function (lines, prefix) {
        var _this = this;
        if (!lines)
            return;
        lines.forEach(function (line) {
            if (line == null)
                return;
            if (_this.maxCommentWidth > 0 && line.length > _this.maxCommentWidth) {
                // See if we can split the line
                var split = CodeWriterUtility.wordWrap(line, _this.maxCommentWidth);
                split.forEach(function (s) {
                    _this.writer.writeLine("" + prefix + s);
                });
            }
            else
                _this.writer.writeLine("" + prefix + line);
        });
    };
    return CSharpCommentWriter;
}());
export { CSharpCommentWriter };
