// @yellicode/angular/dist/es6/angular-writer.js
import { ChangeDetectionStrategy, ViewEncapsulation } from './directive-config';
/**
 * Provides utility functions for generating Angular 2+ application code.
 */
export class AngularWriter {
    /**
    * Writes a full Angular '@Component' class decorator using the provided configuration.
    */
    static writeComponentDecorator(writer, config) {
        writer.writeDecoratorCodeBlock('Component', () => {
            AngularWriter.writeComponentConfig(writer, config);
        });
    }
    /**
     * Writes Angular component configuration that can be used inside a component class decorator.
     * Use writeComponentDecorator() to write a full '@Component' class decorator.
     */
    static writeComponentConfig(writer, config) {
        const keys = Object.keys(config);
        keys.forEach((key, index) => {
            writer.writeIndent();
            const value = config[key];
            writer.write(`${key}: ${AngularWriter.stringifyComponentProperty(key, value)}`);
            if (index < keys.length - 1)
                writer.write(',');
            writer.writeEndOfLine();
        });
    }
    /**
    * Writes a full Angular '@NgModule' class decorator using the provided configuration.
    */
    static writeModuleDecorator(writer, config) {
        writer.writeDecoratorCodeBlock('NgModule', () => {
            AngularWriter.writeComponentConfig(writer, config);
        });
    }
    /**
     * Writes Angular module configuration that can be used inside a module class decorator.
     * Use writeModuleDecorator() to write a full '@NgModule' class decorator.
     */
    static writeModuleConfig(writer, config) {
        const keys = Object.keys(config);
        keys.forEach((key, index) => {
            writer.writeIndent();
            const value = config[key];
            writer.write(`${key}: ${AngularWriter.stringifyModuleProperty(key, value)}`);
            if (index < keys.length - 1)
                writer.write(',');
            writer.writeEndOfLine();
        });
    }
    /**
    * Writes a route configuration for the specified component, with the specified route path.
    */
    static writeRoute(writer, route) {
        writer.writeLine('{');
        writer.increaseIndent();
        writer.writeLine(`path: '${route.path}',`);
        writer.writeLine(`component: ${route.componentName}`);
        writer.decreaseIndent();
        writer.writeLine('},');
    }
    static stringifyModuleProperty(key, value) {
        switch (key) {
            case 'bootstrap':
            case 'declarations':
            case 'entryComponents':
            case 'imports':
            case 'exports':
            case 'providers':
            case 'schemas':
                return AngularWriter.stringifyObjectArray(value);
            default: break;
        }
        if (typeof (value) == typeof (true)) {
            return value ? 'true' : 'false';
        }
        return `'${value}'`;
    }
    static stringifyComponentProperty(key, value) {
        switch (key) {
            case 'changeDetection':
                return `ChangeDetectionStrategy.${ChangeDetectionStrategy[value]}`;
            case 'encapsulation':
                return `ViewEncapsulation.${ViewEncapsulation[value]}`;
            case 'entryComponents':
            case 'providers':
            case 'viewProviders':
                return AngularWriter.stringifyObjectArray(value);
            case 'host':
                return AngularWriter.stringifyKeyValuePair(value);
            default:
                break;
        }
        if (typeof (value) == typeof (true)) {
            return value ? 'true' : 'false';
        }
        if (value instanceof Array) {
            return AngularWriter.stringifyStringArray(value);
        }
        // The value is a string
        else
            return `'${value}'`;
    }
    static stringifyObjectArray(arr) {
        if (!arr.length)
            return '[]';
        return `[${arr.join(', ')}]`;
    }
    static stringifyStringArray(arr) {
        if (!arr.length)
            return '[]';
        return `[${arr.map(item => `'${item}'`).join(', ')}]`;
    }
    static stringifyKeyValuePair(value) {
        // {'key1': 'value1', 'key2': 'value2'}
        const keys = Object.keys(value);
        return `{ ${keys.map(k => `'${k}': '${value[k]}'`).join(', ')} }`;
    }
}
