// @yellicode/elements/dist/es5/document.js
/*
* Copyright (c) 2019 Yellicode
*
* This Source Code Form is subject to the terms of the Mozilla Public
* License, v. 2.0. If a copy of the MPL was not distributed with this
* file, You can obtain one at http://mozilla.org/MPL/2.0/.
*/
var Document = /** @class */ (function () {
    function Document(modelDelegate) {
        this.modelDelegate = modelDelegate;
        this.id = '';
        this.creator = '';
        this.modelTypeName = '';
        this.modelTypeVersion = '';
        this.model = null;
        this.profiles = null;
        this.references = [];
    }
    /**
    * Returns an Element representing the element whose id property matches the specified string.
    * @returns {elements.Element} The model element matching the specified ID, or null if no matching
    * element was found in the model.
    * @param {string} id The ID of the element to search for.
    */
    Document.prototype.findElementById = function (id) {
        return this.modelDelegate.findElementById(id);
    };
    return Document;
}());
export { Document };
