// @yellicode/elements/dist/es5/element-map.js
import { ElementTypeUtility } from './utils';
var ElementMapImpl = /** @class */ (function () {
    function ElementMapImpl() {
        this.elementsById = {};
        this.specializationsById = {};
        this.associationsByEndId = {};
    }
    ElementMapImpl.prototype.addElement = function (element, elementData) {
        if (this.elementsById.hasOwnProperty(element.id)) {
            console.warn("Duplicate element id '" + element.id + "'.");
            return;
        }
        this.elementsById[element.id] = element;
        // Add generalizations to the specialization map
        if (ElementTypeUtility.isClassifier(element.elementType) && elementData) {
            this.addSpecializations(element, elementData);
        }
        // Add association ends to the assciationEnd map
        if (ElementTypeUtility.isAssociation(element.elementType) && elementData) {
            this.addAssociationEnds(element, elementData);
        }
    };
    ElementMapImpl.prototype.addAssociationEnds = function (association, associationData) {
        var _this = this;
        // Get memberEnds of assocationData instead of association itself: the ends  will not be set here as they are not resolved yet
        // by DataToModelConverter.resolveAssociationReferences().        
        if (!associationData.memberEnds)
            return;
        associationData.memberEnds.forEach(function (endId) {
            // An association end can only be part of one association.
            if (_this.associationsByEndId.hasOwnProperty(endId)) {
                console.warn("Association end with id '" + endId + "' is already part of another association than " + associationData.id + ".");
            }
            _this.associationsByEndId[endId] = association;
        });
    };
    ElementMapImpl.prototype.addSpecializations = function (classifier, classifierData) {
        var _this = this;
        if (!classifierData.generalizations)
            return;
        // Enumerate the classifierData instead of the classifier itself: the generalizations will not be set here as they are not resolved yet
        classifierData.generalizations.forEach(function (g) {
            var dictionaryEntry;
            // g is a Generalization of element, so element is a Specialization of g
            if (_this.specializationsById.hasOwnProperty(g.general)) {
                dictionaryEntry = _this.specializationsById[g.general];
                dictionaryEntry.push(classifier);
            }
            else {
                _this.specializationsById[g.general] = [classifier];
            }
        });
    };
    ElementMapImpl.prototype.getAssociationHavingMemberEnd = function (end) {
        if (!end || !end.id)
            return null;
        if (!this.associationsByEndId.hasOwnProperty(end.id))
            return null;
        return this.associationsByEndId[end.id];
    };
    ElementMapImpl.prototype.hasElement = function (id) {
        return this.elementsById.hasOwnProperty(id);
    };
    ElementMapImpl.prototype.getElementById = function (id) {
        if (!id || id.length === 0)
            return null;
        if (this.elementsById.hasOwnProperty(id))
            return this.elementsById[id];
        else {
            console.warn("Unkown element id '" + id + "'.");
            return null;
        }
    };
    ElementMapImpl.prototype.getElementsByIdList = function (idList) {
        var _this = this;
        var result = [];
        if (idList == null)
            return result;
        idList.forEach(function (id) {
            var element = _this.getElementById(id);
            if (element != null)
                result.push(element);
        });
        return result;
    };
    ElementMapImpl.prototype.getSpecializationsOf = function (generalId) {
        if (!this.specializationsById.hasOwnProperty(generalId))
            return [];
        return this.specializationsById[generalId];
    };
    ElementMapImpl.prototype.getAllSpecializationsOf = function (generalId) {
        if (!this.specializationsById.hasOwnProperty(generalId))
            return [];
        var specialMap = {};
        this.getAllSpecializationsRecursive(generalId, specialMap);
        // Convert the result to an array
        var result = [];
        for (var specializationId in specialMap) {
            result.push(specialMap[specializationId]);
        }
        return result;
    };
    ElementMapImpl.prototype.getAllSpecializationsRecursive = function (generalId, specialMap) {
        var _this = this;
        if (!this.specializationsById.hasOwnProperty(generalId)) {
            return;
        }
        var directSpecializations = this.specializationsById[generalId];
        directSpecializations.forEach(function (s) {
            if (!specialMap.hasOwnProperty(s.id)) {
                specialMap[s.id] = s;
            }
            // Get the specializations of this specialization
            _this.getAllSpecializationsRecursive(s.id, specialMap);
        });
    };
    return ElementMapImpl;
}());
export { ElementMapImpl };
