// @yellicode/elements/dist/es6/transforms/dependency-sort-transform.js
/*
* Copyright (c) 2019 Yellicode
*
* This Source Code Form is subject to the terms of the Mozilla Public
* License, v. 2.0. If a copy of the MPL was not distributed with this
* file, You can obtain one at http://mozilla.org/MPL/2.0/.
*/
import * as toposort from 'toposort';
import { isPackage, isClassifier, isBehavioredClassifier, isMemberedClassifier, isType } from '../utils';
/**
 * A bitwise enumeration used to specify what types of dependency to include
 * during a DependencySortTransform.
 */
export var DependencyKind;
(function (DependencyKind) {
    DependencyKind[DependencyKind["none"] = 0] = "none";
    DependencyKind[DependencyKind["generalizations"] = 1] = "generalizations";
    DependencyKind[DependencyKind["interfaceRealizations"] = 2] = "interfaceRealizations";
    DependencyKind[DependencyKind["attributes"] = 4] = "attributes";
    DependencyKind[DependencyKind["operationParameters"] = 8] = "operationParameters";
    DependencyKind[DependencyKind["all"] = 15] = "all";
})(DependencyKind || (DependencyKind = {}));
/**
 * Sorts the types within a package (and within each nested package) based on their dependencies, in such a way
 * that dependencies appear before dependents. Dependencies are determined based on generalizations, interface
 * realizations, attributes and operation parameters.
 */
export class DependencySortTransform {
    /**
     * Constructor. Creates a new DependencySortTransform instance.
     * @param dependencyKind An optional DependencyKind value (or a bitwise combination of values)
     * that indicates what types of dependency must be taken into account. The default is DependencyKind.All.
     */
    constructor(dependencyKind) {
        this._options = dependencyKind == null ? DependencyKind.all : dependencyKind;
    }
    transform(pack) {
        if (!pack)
            return pack;
        this.transformPackageRecursive(pack);
        return pack;
    }
    transformPackageRecursive(pack) {
        if (!pack.packagedElements)
            return;
        // Get all siblings that are relevant for building a depencency graph.        
        const allTypes = pack.packagedElements.filter(pe => isType(pe));
        // Build a dependency graph of each element, see https://www.npmjs.com/package/toposort for the docs
        var graph = [];
        allTypes.forEach((t) => {
            let elementDependencies = DependencySortTransform.getTypeDependencies(t, this._options, allTypes);
            if (elementDependencies.length > 0) {
                elementDependencies.forEach((dependency) => {
                    graph.push([t, dependency]);
                });
            }
        });
        if (graph.length > 0) {
            // Sort, and reverse because we need a dependency graph
            pack.packagedElements = toposort.array(pack.packagedElements, graph).reverse();
        }
        // Transform nested packages, recursively        
        pack.packagedElements.forEach((element) => {
            if (isPackage(element)) {
                this.transformPackageRecursive(element);
            }
        });
    }
    static getTypeDependencies(element, options, allTypes) {
        var result = [];
        // Dependencies based on generalizations
        if (options & DependencyKind.generalizations && isClassifier(element)) {
            DependencySortTransform.pushGeneralizationDependencies(element, allTypes, result);
        }
        // Dependencies based on interface realization
        if (options & DependencyKind.interfaceRealizations && isBehavioredClassifier(element)) {
            DependencySortTransform.pushInterfaceRealizationDependencies(element, allTypes, result);
        }
        // Dependencies based on members
        if (isMemberedClassifier(element)) {
            if (options & DependencyKind.attributes) {
                DependencySortTransform.pushOwnedAttributeDependencies(element, allTypes, result);
            }
            if (options & DependencyKind.operationParameters) {
                DependencySortTransform.pushOwnedOperationDependencies(element, allTypes, result);
            }
        }
        return result;
    }
    static pushGeneralizationDependencies(element, allTypes, dependencies) {
        if (!element.generalizations || !element.generalizations.length)
            return;
        element.generalizations.forEach(g => {
            if (g.general === element)
                return;
            if (allTypes.indexOf(g.general) > -1 && dependencies.indexOf(g.general) === -1) {
                dependencies.push(g.general);
            }
        });
    }
    static pushInterfaceRealizationDependencies(element, allTypes, dependencies) {
        if (!element.interfaceRealizations || !element.interfaceRealizations.length)
            return;
        element.interfaceRealizations.forEach(ir => {
            if (allTypes.indexOf(ir.contract) > -1 && dependencies.indexOf(ir.contract) === -1) {
                dependencies.push(ir.contract);
            }
        });
    }
    static pushOwnedOperationDependencies(element, allTypes, dependencies) {
        if (!element.ownedOperations || !element.ownedOperations.length)
            return;
        element.ownedOperations.forEach(op => {
            if (!op.ownedParameters || !op.ownedParameters.length)
                return;
            op.ownedParameters.forEach(p => {
                if (!p.type)
                    return;
                if (allTypes.indexOf(p.type) > -1 && dependencies.indexOf(p.type) === -1) {
                    dependencies.push(p.type);
                }
            });
        });
    }
    static pushOwnedAttributeDependencies(element, allTypes, dependencies) {
        if (!element.ownedAttributes || !element.ownedAttributes.length)
            return;
        element.ownedAttributes.forEach(att => {
            if (!att.type)
                return;
            if (allTypes.indexOf(att.type) > -1 && dependencies.indexOf(att.type) === -1) {
                dependencies.push(att.type);
            }
        });
    }
}
