// @yellicode/templating/dist/es5/code-writer.js
/**
 * The abstract base class for all code writers. This writer implements the TextWriter interface by decorating
 * the TextWriter provided in the constructor.
 * @deprecated CodeWriter has moved to the 'yellicode/core' package, please update your references.
 */
var CodeWriter = /** @class */ (function () {
    /**
     * Constructor. Creates a new CodeWriter that uses the provided TextWriter internally.
     * @param writer A TextWriter object. In a code generation template, a TextWriter instance can be obtained by calling any of the generate...()
     * functions on the current CodeGenerator.
     */
    function CodeWriter(writer) {
        console.warn("CodeWriter has moved to the '@yellicode/core' package, please update your references.");
        this.writer = writer;
    }
    Object.defineProperty(CodeWriter.prototype, "endOfLineString", {
        // NOTE: only documenting the get accessors because the TS compiler we use for doc generation concats the doc 
        // comments for the getter and setter.
        /**
        * Gets or sets the end of line string. The default value is platform dependent.
        */
        get: function () {
            return this.writer.endOfLineString;
        },
        set: function (value) {
            this.writer.endOfLineString = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CodeWriter.prototype, "indentString", {
        /**
         * Gets or sets the indent string. The default value is a '\t' (tab character).
        */
        get: function () {
            return this.writer.indentString;
        },
        set: function (value) {
            this.writer.indentString = value;
        },
        enumerable: true,
        configurable: true
    });
    /******************************************************************************
     *                          TextWriter delegation
    ******************************************************************************/
    /**
     * Writes a string value to the output.
     * @param value The string value to be written.
     */
    CodeWriter.prototype.write = function (value) {
        this.writer.write(value);
        return this;
    };
    ;
    /**
    * Writes a single whitespace character to the output.
    */
    CodeWriter.prototype.writeWhiteSpace = function () {
        this.writer.writeWhiteSpace();
        return this;
    };
    /**
    * Writes a new line to the output. The line is indented automatically. The line is ended with the endOfLineString.
    * @param value The line to write. When omitted, only the endOfLineString is written.
    */
    CodeWriter.prototype.writeLine = function (value) {
        this.writer.writeLine(value);
        return this;
    };
    ;
    /**
   * Writes a collection of lines to the output. Each line is indented automatically and ended with the endOfLineString.
   * @param values The lines to write.
   * @param delimiter An optional delimiter to be written at the end of each line, except for the last one.
   */
    CodeWriter.prototype.writeLines = function (values, delimiter) {
        this.writer.writeLines(values, delimiter);
        return this;
    };
    /**
     * Writes a new line to the output while temporarily increasing the indent. The line is ended with the endOfLineString.
     * @param value The line to write.
     */
    CodeWriter.prototype.writeLineIndented = function (value) {
        this.writer.writeLineIndented(value);
        return this;
    };
    /**
     * Writes the endOfLineString to the output.
     * @param value Any value to write before the endOfLineString string is written.
     */
    CodeWriter.prototype.writeEndOfLine = function (value) {
        this.writer.writeEndOfLine(value);
        return this;
    };
    ;
    /**
      * Writes the contents of the specified file to the output.
      * @param fileName The path of the file, relative to the template.
      * @param encoding Optional: the encoding that is used for the file. The default is 'utf-8'.
      */
    CodeWriter.prototype.writeFile = function (path, encoding) {
        this.writer.writeFile(path, encoding);
        return this;
    };
    /**
    * Writes the contents of the specified file region to the output.
    * @param regionName The name of the region to write. The region should be marked using "/// <region>code goes here...</region>".
    * @param fileName The path of the file, relative to the template.
    * @param encoding The encoding that is used for the file. The default is 'utf-8'.
    */
    CodeWriter.prototype.writeFileRegion = function (regionName, path, encoding) {
        return this.writer.writeFileRegion(regionName, path, encoding);
    };
    /**
    * Writes the current indentString to the output.
    */
    CodeWriter.prototype.writeIndent = function () {
        this.writer.writeIndent();
        return this;
    };
    ;
    /**
    * Increases the current indent, which is prefixed to each line of the output.
    */
    CodeWriter.prototype.increaseIndent = function () {
        this.writer.increaseIndent();
        return this;
    };
    /**
    * Decreases the current indent, which is prefixed to each line of the output.
    */
    CodeWriter.prototype.decreaseIndent = function () {
        this.writer.decreaseIndent();
        return this;
    };
    /**
     * Resets any indentation, causing new line writes to start at the first character position.
     */
    CodeWriter.prototype.clearIndent = function () {
        this.writer.clearIndent();
        return this;
    };
    return CodeWriter;
}());
export { CodeWriter };
