// @yellicode/templating/dist/es5/generator.js
/*
 * Copyright (c) 2019 Yellicode
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import * as path from 'path';
import * as fs from 'fs';
import { ModelReader } from '@yellicode/elements';
import { StreamWriter } from './stream-writer';
import { FileSystemUtility } from './file-system-utility';
export var OutputMode;
(function (OutputMode) {
    /**
     * The output file will be truncated if it exists. This is the default value.
     */
    OutputMode[OutputMode["Overwrite"] = 0] = "Overwrite";
    /**
     * The output file will not be truncated if it already exists. Use this option if you want
     * to update the file manually once it is generated.
     */
    OutputMode[OutputMode["Once"] = 1] = "Once";
    /**
     * The template output will be appended to the file if it already exists. The file is created if it
     * does not exist.
     */
    OutputMode[OutputMode["Append"] = 2] = "Append";
})(OutputMode || (OutputMode = {}));
var InternalGenerator = /** @class */ (function () {
    function InternalGenerator() {
        this.outputMode = OutputMode.Overwrite;
        //this.templateArgs = InternalGenerator.parseTemplateArgs(process.argv);
        this.parseProcessArgs(process.argv);
        var startedMessage = { cmd: 'processStarted' };
        this.sendProcessMessage(startedMessage);
    }
    InternalGenerator.prototype.parseProcessArgs = function (args) {
        for (var index = 0; index < args.length; index++) {
            var val = args[index];
            if (val === '--templateArgs') {
                this.templateArgs = InternalGenerator.parseTemplateArgs(args, index);
            }
            else if (val === '--outputMode') {
                this.outputMode = InternalGenerator.parseOutputMode(args, index) || this.outputMode;
            }
        }
    };
    InternalGenerator.parseOutputMode = function (args, index) {
        if (args.length <= index + 1)
            return null;
        var outputModeString = args[index + 1];
        switch (outputModeString) {
            case 'append':
                return OutputMode.Append;
            case 'once':
                return OutputMode.Once;
            case 'overwrite':
                return OutputMode.Overwrite;
            default:
                return null;
        }
    };
    InternalGenerator.parseTemplateArgs = function (args, index) {
        if (args.length <= index + 1)
            return null;
        var templateArgsString = args[index + 1];
        if (templateArgsString.length > 0) {
            return JSON.parse(templateArgsString);
        }
        return null;
    };
    InternalGenerator.prototype.sendProcessMessage = function (message) {
        process.send(message); // https://github.com/Microsoft/TypeScript/issues/10158        
    };
    InternalGenerator.prototype.generate = function (options, template) {
        this.generateInternal(options, function (writer) {
            template(writer);
            return Promise.resolve();
        });
    };
    InternalGenerator.prototype.generateAsync = function (options, template) {
        this.generateInternal(options, template);
    };
    /**
     * Executes the provided template using the model that was configured in the code generation configuration.
     */
    InternalGenerator.prototype.generateFromModel = function (options, template) {
        var _this = this;
        this.getModel(options)
            .then(function (model) {
            _this.generateInternal(options, function (writer) {
                template(writer, model);
                return Promise.resolve();
            });
        }).catch(function (err) {
            console.log(err);
        });
    };
    InternalGenerator.prototype.generateFromModelAsync = function (options, template) {
        var _this = this;
        this.getModel(options)
            .then(function (model) {
            _this.generateInternal(options, function (writer) {
                return template(writer, model);
            });
        }).catch(function (err) {
            console.log(err);
        });
    };
    InternalGenerator.prototype.generateInternal = function (options, callback) {
        var _this = this;
        // Get the working directory. The host will make sure that this is the directory in which the template resides.
        var templateDirName = path.resolve('./');
        // console.log('Generator: Template directory name is \'%s\'', templateDirName);
        var fullOutputFileName = path.join(templateDirName, options.outputFile);
        // Ensure that the directory exists        
        FileSystemUtility.ensureDirectory(path.dirname(fullOutputFileName));
        // Let the host know that we started something so that we don't get killed
        var startedMessage = { cmd: 'generateStarted' };
        this.sendProcessMessage(startedMessage);
        // console.log('Generator: Generating file \'%s\'...', fullOutputFileName);
        var mode = options.outputMode === undefined ? this.outputMode : options.outputMode;
        if (mode === OutputMode.Once && fs.existsSync(fullOutputFileName)) {
            // Don't regenerate the file
            var finishedMessage = { cmd: 'generateFinished' };
            this.sendProcessMessage(finishedMessage);
            return;
        }
        var flags = mode === OutputMode.Append ? 'a' : 'w';
        var writeStream = fs.createWriteStream(fullOutputFileName, { flags: flags });
        writeStream.once('open', function (fd) {
            var cw = new StreamWriter(writeStream, options.regionMarkerFormatter);
            callback(cw)
                .then(function () {
                writeStream.end();
                // console.log('Generator: Finished generating \'%s\'...', fullOutputFileName);
                // Let the host know that we are done
                var finishedMessage = { cmd: 'generateFinished' };
                _this.sendProcessMessage(finishedMessage);
            });
        });
    };
    /**
    * Loads the model that was configured in the code generation configuration.
    */
    InternalGenerator.prototype.getModel = function (options) {
        var codeModelOptions = options || {};
        var parseJson = codeModelOptions.noParse !== true;
        var promise = new Promise(function (resolve, reject) {
            process.on('message', function (m) {
                if (m.cmd !== 'setModel') {
                    return;
                }
                if (!m.modelData) {
                    return reject('The host returned an empty model. Please make sure that a model has been configured for this template.');
                }
                var model;
                // Should we parse the model into a Yellicode model?                                
                if (parseJson && ModelReader.canRead(m.modelData)) {
                    // The modelData will contain a Yellicode document with two main nodes: a 'model' node and an optional 'profiles' node.
                    // We need to parse the entire document (because profiles must be applied) and then return
                    // just the model part.
                    var document_1 = ModelReader.readDocument(m.modelData);
                    if (document_1) {
                        model = document_1.model;
                    }
                    else
                        model = null;
                }
                else
                    model = m.modelData; // return plain JSON
                // Apply transforms                
                var targetModel;
                if (model && codeModelOptions.modelTransform) {
                    targetModel = codeModelOptions.modelTransform.transform(model);
                }
                else
                    targetModel = model;
                resolve(targetModel);
            });
        });
        var getModelMessage = { cmd: 'getModel' };
        this.sendProcessMessage(getModelMessage);
        return promise;
    };
    return InternalGenerator;
}());
export var Generator = new InternalGenerator();
