// @yellicode/templating/dist/es5/stream-writer.js
/*
 * Copyright (c) 2019 Yellicode
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
var fs = require('fs');
var path = require('path');
var os = require('os');
import { DefaultRegionMarkerFormatter } from './region-marker-formatter';
/**
 * Internal writer class that is passed to WriterBase implementations. We don't let people inherit from TextWriter directly because
 * we want to control the instantiation and lifetime of each StreamWriter instance.
 */
var StreamWriter = /** @class */ (function () {
    /**
     * Constructor. Creates a new StreamWriter that writes to the provided WritableStream.
     * @param stream A Node.js WritableStream instance.
     * @param regionMarkerFormatter An optional RegionMarkerFormatter.
     */
    function StreamWriter(stream, regionMarkerFormatter) {
        this.stream = stream;
        this.indent = 0;
        this.textFileCache = {};
        this.endOfLineString = os.EOL;
        this.indentString = "\t";
        this.templateDirName = path.resolve('.');
        this.fileRegionMapper = regionMarkerFormatter == null ? new DefaultRegionMarkerFormatter() : regionMarkerFormatter;
    }
    /******************************************************************************
     *                          TextWriter implementation
    ******************************************************************************/
    /**
    * Writes a new line to the output. The line is indented automatically. The line is ended with the endOfLineString.
    * @param value The line to write. When omitted, only the endOfLineString is written.
    */
    StreamWriter.prototype.writeLine = function (value) {
        if (value) {
            this.stream.write(this.createIndentString() + value + this.endOfLineString);
        }
        else
            this.stream.write(this.endOfLineString);
        return this;
    };
    ;
    /**
    * Writes a collection of lines to the output. Each line is indented automatically and ended with the endOfLineString.
    * @param values The lines to write.
    * @param delimiter An optional delimiter to be written at the end of each line, except for the last one.
    */
    StreamWriter.prototype.writeLines = function (values, delimiter) {
        var _this = this;
        if (!values)
            return this;
        var len = values.length;
        values.forEach(function (value, index) {
            _this.stream.write(_this.createIndentString() + value);
            if (delimiter && index < len - 1) {
                _this.stream.write(delimiter);
            }
            _this.stream.write(_this.endOfLineString);
        });
        return this;
    };
    /**
     * Writes a new line to the output while temporarily increasing the indent. The line is ended with the endOfLineString.
     * @param value The line to write.
     */
    StreamWriter.prototype.writeLineIndented = function (value) {
        this.stream.write(this.indentString + this.createIndentString() + value + this.endOfLineString);
        return this;
    };
    StreamWriter.prototype.writeEndOfLine = function (value) {
        if (value) {
            this.stream.write(value);
        }
        this.stream.write(this.endOfLineString);
        return this;
    };
    StreamWriter.prototype.writeIndent = function () {
        this.stream.write(this.createIndentString());
        return this;
    };
    ;
    StreamWriter.prototype.write = function (value) {
        if (value == null)
            return this; // avoid 'May not write null values to stream'
        this.stream.write(value);
        return this;
    };
    ;
    StreamWriter.prototype.writeWhiteSpace = function () {
        this.write(" ");
        return this;
    };
    StreamWriter.prototype.increaseIndent = function () {
        this.indent++;
        return this;
    };
    StreamWriter.prototype.decreaseIndent = function () {
        if (this.indent > 0) {
            this.indent--;
        }
        return this;
    };
    StreamWriter.prototype.clearIndent = function () {
        this.indent = 0;
        return this;
    };
    StreamWriter.prototype.writeFile = function (path, encoding) {
        var fullPath = this.resolveFileName(path);
        var contents = this.readTextFile(fullPath, false, encoding);
        if (contents == null || contents.length <= 0)
            return this;
        this.stream.write(contents);
        return this;
    };
    StreamWriter.prototype.writeFileRegion = function (regionName, path, encoding) {
        if (this.fileRegionMapper == null || regionName == null || regionName.length === 0 || path == null || path.length === 0)
            return false;
        var fullPath = this.resolveFileName(path);
        var contents = this.readTextFile(fullPath, true, encoding);
        if (contents == null || contents.length <= 0)
            return false;
        // First find the region match 
        var regionStartMarker = this.fileRegionMapper.getRegionStartMarker(regionName);
        var regionStartIndex = contents.indexOf(regionStartMarker);
        if (regionStartIndex < 0) {
            return false;
        }
        var regionEndMarker = this.fileRegionMapper.getRegionEndMarker(regionName);
        var regionEndIndex = contents.indexOf(regionEndMarker, regionStartIndex);
        if (regionEndIndex < 0) {
            return false;
        }
        var region = contents.substring(regionStartIndex + regionStartMarker.length, regionEndIndex);
        this.stream.write(region);
        this.writeEndOfLine();
        return true;
    };
    /******************************************************************************
     *                         End of TextWriter implementation
    ******************************************************************************/
    StreamWriter.prototype.createIndentString = function () {
        var result = "";
        for (var i = 0; i < this.indent; i++) {
            result += this.indentString;
        }
        return result;
    };
    ;
    StreamWriter.prototype.resolveFileName = function (fileName) {
        return path.join(this.templateDirName, fileName);
    };
    StreamWriter.prototype.readTextFile = function (path, useCache, encoding) {
        if (useCache && this.textFileCache.hasOwnProperty(path)) {
            return this.textFileCache[path];
        }
        if (!fs.existsSync(path)) {
            //  this.writeLine(`Cannot write the contents of file '${fullPath}' because the file does not exist.`);
            if (useCache) {
                this.textFileCache[path] = null;
            }
            return undefined;
        }
        if (encoding != null) {
            encoding = encoding.toLowerCase();
        }
        else
            encoding = 'utf-8';
        var contents = fs.readFileSync(path, encoding);
        if (contents == null || contents.length === 0)
            return undefined;
        // Remove any utf-8 BOM if there is any
        if (encoding === 'utf-8' && contents.charCodeAt(0) === 0xFEFF) {
            contents = contents.slice(1);
        }
        if (useCache) {
            this.textFileCache[path] = contents;
        }
        return contents;
    };
    return StreamWriter;
}());
export { StreamWriter };
