// @yellicode/templating/dist/es5/to-host-logger.js
/*
 * Copyright (c) 2019 Yellicode
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import { LogLevel } from '@yellicode/core';
/**
 * Uses the active host process logger for log message.
 */
var ToHostLogger = /** @class */ (function () {
    function ToHostLogger() {
    }
    ToHostLogger.prototype.log = function (message, level) {
        var processMessage = { log: { level: level, message: message } };
        process.send(processMessage); // https://github.com/Microsoft/TypeScript/issues/10158
    };
    ToHostLogger.prototype.verbose = function (message) {
        this.log(message, LogLevel.Verbose);
    };
    ToHostLogger.prototype.info = function (message) {
        this.log(message, LogLevel.Info);
    };
    ToHostLogger.prototype.warn = function (message) {
        this.log(message, LogLevel.Warning);
    };
    ToHostLogger.prototype.error = function (message) {
        this.log(message, LogLevel.Error);
    };
    return ToHostLogger;
}());
export { ToHostLogger };
