// @yellicode/elements/dist/es6/element-map.js
import { ElementTypeUtility } from './utils';
export class ElementMapImpl {
    constructor() {
        this.elementsById = {};
        this.specializationsById = {};
        this.associationsByEndId = {};
    }
    addElement(element, elementData) {
        if (this.elementsById.hasOwnProperty(element.id)) {
            console.warn(`Duplicate element id '${element.id}'.`);
            return;
        }
        this.elementsById[element.id] = element;
        // Add generalizations to the specialization map
        if (ElementTypeUtility.isClassifier(element.elementType) && elementData) {
            this.addSpecializations(element, elementData);
        }
        // Add association ends to the assciationEnd map
        if (ElementTypeUtility.isAssociation(element.elementType) && elementData) {
            this.addAssociationEnds(element, elementData);
        }
    }
    addAssociationEnds(association, associationData) {
        // Get memberEnds of assocationData instead of association itself: the ends  will not be set here as they are not resolved yet
        // by DataToModelConverter.resolveAssociationReferences().        
        if (!associationData.memberEnds)
            return;
        associationData.memberEnds.forEach(endId => {
            // An association end can only be part of one association.
            if (this.associationsByEndId.hasOwnProperty(endId)) {
                console.warn(`Association end with id '${endId}' is already part of another association than ${associationData.id}.`);
            }
            this.associationsByEndId[endId] = association;
        });
    }
    addSpecializations(classifier, classifierData) {
        if (!classifierData.generalizations)
            return;
        // Enumerate the classifierData instead of the classifier itself: the generalizations will not be set here as they are not resolved yet
        classifierData.generalizations.forEach(g => {
            let dictionaryEntry;
            // g is a Generalization of element, so element is a Specialization of g
            if (this.specializationsById.hasOwnProperty(g.general)) {
                dictionaryEntry = this.specializationsById[g.general];
                dictionaryEntry.push(classifier);
            }
            else {
                this.specializationsById[g.general] = [classifier];
            }
        });
    }
    getAssociationHavingMemberEnd(end) {
        if (!end || !end.id)
            return null;
        if (!this.associationsByEndId.hasOwnProperty(end.id))
            return null;
        return this.associationsByEndId[end.id];
    }
    hasElement(id) {
        return this.elementsById.hasOwnProperty(id);
    }
    getElementById(id) {
        if (!id || id.length === 0)
            return null;
        if (this.elementsById.hasOwnProperty(id))
            return this.elementsById[id];
        else {
            console.warn(`Unkown element id '${id}'.`);
            return null;
        }
    }
    getElementsByIdList(idList) {
        const result = [];
        if (idList == null)
            return result;
        idList.forEach((id) => {
            const element = this.getElementById(id);
            if (element != null)
                result.push(element);
        });
        return result;
    }
    getSpecializationsOf(generalId) {
        if (!this.specializationsById.hasOwnProperty(generalId))
            return [];
        return this.specializationsById[generalId];
    }
    getAllSpecializationsOf(generalId) {
        if (!this.specializationsById.hasOwnProperty(generalId))
            return [];
        const specialMap = {};
        this.getAllSpecializationsRecursive(generalId, specialMap);
        // Convert the result to an array
        const result = [];
        for (var specializationId in specialMap) {
            result.push(specialMap[specializationId]);
        }
        return result;
    }
    getAllSpecializationsRecursive(generalId, specialMap) {
        if (!this.specializationsById.hasOwnProperty(generalId)) {
            return;
        }
        const directSpecializations = this.specializationsById[generalId];
        directSpecializations.forEach(s => {
            if (!specialMap.hasOwnProperty(s.id)) {
                specialMap[s.id] = s;
            }
            // Get the specializations of this specialization
            this.getAllSpecializationsRecursive(s.id, specialMap);
        });
    }
}
