// @yellicode/elements/dist/es6/model-delegate.js
/*
 * Copyright (c) 2019 Yellicode
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import * as Interfaces from "./interfaces";
import { ElementTypeUtility } from './utils';
import * as _ from 'lodash';
import { ElementComparerImpl } from "./element-comparer";
import * as utils from './utils';
/**
 * Internal class to which all behaviour of the model classes is delegated.
 */
export class ModelDelegateImpl {
    constructor(elementMap) {
        this.elementMap = elementMap;
        // Note that elementMap is null when not initialized through the DataToModelConverter
    }
    // ************************  Document **************************** //     
    findElementById(id) {
        if (!this.elementMap)
            return null;
        return this.elementMap.getElementById(id);
    }
    // **************************  Element ******************************* //     
    getFirstCommentBody(element) {
        if (!element.ownedComments || element.ownedComments.length == 0)
            return ''; // return an empty string, no null or undefined, they are inconvenient when using template literals.      
        return element.ownedComments[0].body;
    }
    // ************************  TypedElement **************************** //     
    getTypeName(typedElement) {
        if (!typedElement.type)
            return '';
        return typedElement.type.name;
    }
    // *****************************  Package **************************** //     
    static getPackagedElementsWhere(pack, predicate) {
        if (!pack.packagedElements)
            return [];
        return pack.packagedElements.filter(predicate);
    }
    static getAllPackagedElementsWhereRecursive(pack, predicate, result) {
        pack.packagedElements.forEach(e => {
            if (predicate(e)) {
                result.push(e);
            }
            if (ElementTypeUtility.isPackage(e.elementType) && e.packagedElements) {
                ModelDelegateImpl.getAllPackagedElementsWhereRecursive(e, predicate, result);
            }
        });
    }
    static getAllPackagedElementsWhere(pack, predicate) {
        const result = [];
        if (pack.packagedElements) {
            ModelDelegateImpl.getAllPackagedElementsWhereRecursive(pack, predicate, result);
        }
        return result;
    }
    getNestedPackages(element) {
        return ModelDelegateImpl.getPackagedElementsWhere(element, pe => pe.elementType === Interfaces.ElementType.package);
    }
    getTypes(element) {
        return ModelDelegateImpl.getPackagedElementsWhere(element, pe => 
        // Technically, an Association is also a Type, but it does not seem to make much sense here
        ElementTypeUtility.isType(pe.elementType) && !ElementTypeUtility.isAssociation(pe.elementType));
    }
    getAllTypes(element) {
        return ModelDelegateImpl.getAllPackagedElementsWhere(element, pe => 
        // Technically, an Association is also a Type, but it does not seem to make much sense here
        ElementTypeUtility.isType(pe.elementType) && !ElementTypeUtility.isAssociation(pe.elementType));
    }
    getClasses(element) {
        return ModelDelegateImpl.getPackagedElementsWhere(element, pe => ElementTypeUtility.isClass(pe.elementType));
    }
    getAllClasses(element) {
        return ModelDelegateImpl.getAllPackagedElementsWhere(element, pe => ElementTypeUtility.isClass(pe.elementType));
    }
    getInterfaces(element) {
        return ModelDelegateImpl.getPackagedElementsWhere(element, pe => ElementTypeUtility.isInterface(pe.elementType));
    }
    getAllInterfaces(element) {
        return ModelDelegateImpl.getAllPackagedElementsWhere(element, pe => ElementTypeUtility.isInterface(pe.elementType));
    }
    getDataTypes(element) {
        return ModelDelegateImpl.getPackagedElementsWhere(element, pe => ElementTypeUtility.isDataType(pe.elementType));
    }
    getAllDataTypes(element) {
        return ModelDelegateImpl.getAllPackagedElementsWhere(element, pe => ElementTypeUtility.isDataType(pe.elementType));
    }
    getEnumerations(element) {
        return ModelDelegateImpl.getPackagedElementsWhere(element, pe => ElementTypeUtility.isEnumeration(pe.elementType));
    }
    getAllEnumerations(element) {
        return ModelDelegateImpl.getAllPackagedElementsWhere(element, pe => ElementTypeUtility.isEnumeration(pe.elementType));
    }
    // ************************  PackageableElement **************************** //     
    getPackage(packagedElement) {
        return packagedElement.owner;
    }
    /**
    * Gets all packages that own this Package, working inwards from the top Package to the owning package.
    * @returns {Interfaces.Package[]} A collection of Packages.
    */
    getNestingPackages(packagedElement, stopAtNamespaceRoot) {
        // Work outwards, then reverse..
        const result = [];
        let owner = packagedElement.owner;
        while (owner && utils.isPackage(owner)) {
            result.push(owner);
            if (owner.isNamespaceRoot && stopAtNamespaceRoot === true) {
                break;
            }
            owner = owner.owner;
        }
        result.reverse();
        return result;
    }
    getNamespaceName(packagedElement, separator = '.') {
        return this.getNestingPackages(packagedElement, true).map(p => p.name).join(separator);
    }
    getQualifiedName(packagedElement, separator = '.') {
        const namespaceName = this.getNamespaceName(packagedElement, separator);
        return namespaceName ? `${namespaceName}${separator}${packagedElement.name}` : packagedElement.name;
    }
    // ************************  Generalization **************************** //     
    getSpecific(generalization) {
        return generalization.owner;
    }
    // **************************  Classifier ******************************* //     
    getFirstGeneralization(classifier) {
        if (classifier.generalizations.length === 0)
            return null;
        return classifier.generalizations[0];
    }
    getFirstParent(classifier) {
        const firstGeneralization = this.getFirstGeneralization(classifier);
        return (firstGeneralization) ? firstGeneralization.general : null;
    }
    getParents(classifier) {
        return classifier.generalizations.map(g => g.general);
    }
    /**
    * Returns all of the direct and indirect ancestors of a generalized Classifier, working outwards: more specific classifiers will
    * appear before more general classifiers.
    */
    getAllParents(classifier) {
        const allParents = [];
        this.getAllParentsRecursive(classifier, allParents);
        return allParents;
    }
    getAllParentsRecursive(classifier, allParents) {
        const ownParents = this.getParents(classifier);
        if (ownParents != null) {
            // More specific parents first.
            ownParents.forEach(p => {
                const ix = allParents.indexOf(p);
                // If the parent is already there, remove it and add it again at the end of the list.
                if (ix > -1) {
                    allParents.splice(ix, 1);
                }
                allParents.push(p);
            });
            // add the parents of each parent
            ownParents.forEach(p => {
                this.getAllParentsRecursive(p, allParents);
            });
        }
    }
    getSpecializations(classifier) {
        if (!this.elementMap)
            return [];
        return this.elementMap.getSpecializationsOf(classifier.id);
    }
    getAllSpecializations(classifier) {
        if (!this.elementMap)
            return [];
        return this.elementMap.getAllSpecializationsOf(classifier.id);
    }
    // /**
    //  * Returns true if one classifier has the other as direct general.     
    //  */
    // private static hasGeneral(x: Interfaces.Classifier, y: Interfaces.Classifier): boolean {
    //     // TODO: recursive?
    //     const firstMatch = x.generalizations.find(g => g.general.id === y.id);
    //     return firstMatch ? true: false;    
    // }
    // ********************  MemberedClassifier **************************** //     
    getAllAttributes(memberedClassifier) {
        const distinctClassifierIds = []; // keep this list to check against recursion
        const result = [];
        this.getAllAttributesRecursive(memberedClassifier, distinctClassifierIds, result);
        return result;
    }
    getAllAttributesRecursive(memberedClassifier, distinctClassifierIds, result) {
        distinctClassifierIds.push(memberedClassifier.id);
        // First add the general members, then the owned
        memberedClassifier.generalizations.forEach(g => {
            const general = g.general;
            if (distinctClassifierIds.indexOf(general.id) === -1) {
                this.getAllAttributesRecursive(general, distinctClassifierIds, result);
            }
        });
        // Now add the owned members
        result.push(...memberedClassifier.ownedAttributes);
    }
    getAllOperations(memberedClassifier) {
        const distinctClassifierIds = []; // keep this list to check against recursion
        const result = [];
        this.getAllOperationsRecursive(memberedClassifier, distinctClassifierIds, result);
        return result;
    }
    getAllOperationsRecursive(memberedClassifier, distinctClassifierIds, result) {
        distinctClassifierIds.push(memberedClassifier.id); // keep this list to check against recursion
        // First add the general members, then the owned
        memberedClassifier.generalizations.forEach(g => {
            const general = g.general;
            if (distinctClassifierIds.indexOf(general.id) === -1) {
                this.getAllOperationsRecursive(general, distinctClassifierIds, result);
            }
        });
        // Now add the owned members. Replace any inherited members with the same signature
        memberedClassifier.ownedOperations.forEach(op => {
            // Is there an operation in the result that has the same signature? Then replace it.
            // TODO: can we set a "redefines" reference (referring to the base operation)?             
            _.remove(result, (baseOperation) => ElementComparerImpl.haveEqualSignatures(baseOperation, op));
            result.push(op);
        });
    }
    // ************************  Class  **************************** //         
    getSuperClasses(cls) {
        const result = [];
        if (!cls.generalizations)
            return result;
        cls.generalizations.forEach(g => {
            if (ElementTypeUtility.isClass(g.general.elementType)) {
                result.push(g.general);
            }
        });
        return result;
    }
    // ************************  MultiplicityElement **************************** //         
    getLower(element) {
        // If element is an operation, return the lower of the return parameter
        const multiplicityElement = element.elementType === Interfaces.ElementType.operation ?
            element.getReturnParameter() :
            element;
        if (!multiplicityElement)
            return null; // the operation has no return parameter
        const lowerValue = multiplicityElement.lowerValue;
        if (!lowerValue)
            return null;
        switch (lowerValue.elementType) {
            case Interfaces.ElementType.literalInteger:
                return lowerValue.value;
            case Interfaces.ElementType.literalString:
                const stringValue = lowerValue.getStringValue();
                const parsed = Number.parseInt(stringValue);
                return isNaN(parsed) ? null : parsed;
            default:
                return null;
        }
    }
    getUpper(element) {
        // If element is an operation, return the upper of the return parameter
        const multiplicityElement = element.elementType === Interfaces.ElementType.operation ?
            element.getReturnParameter() :
            element;
        if (!multiplicityElement)
            return null; // the operation has no return parameter
        const upperValue = multiplicityElement.upperValue;
        if (upperValue == null)
            return null;
        switch (upperValue.elementType) {
            case Interfaces.ElementType.literalString:
            case Interfaces.ElementType.literalInteger:
                const stringValue = upperValue.getStringValue();
                return new Interfaces.UnlimitedNatural(stringValue);
            case Interfaces.ElementType.literalUnlimitedNatural:
                return upperValue.value;
            default:
                return null;
        }
    }
    getLowerBound(element) {
        const lower = this.getLower(element);
        // Do an explicit null check, allowing for 0. An unspecified lower means a lower bound of 1.
        return (lower === null) ? 1 : lower;
    }
    getUpperBound(element) {
        const upper = this.getUpper(element);
        // Do an explicit null check, allowing for 0. An unspecified upper means an upper bound of 1.
        return (upper === null) ? new Interfaces.UnlimitedNatural(1) : upper;
    }
    isMultivalued(element) {
        const upper = this.getUpper(element);
        if (!upper)
            return false;
        return upper.IsInfinity || upper.Value > 1;
    }
    isOptional(element) {
        return this.getLowerBound(element) === 0;
    }
    isOptionalAndSinglevalued(element) {
        if (this.getLowerBound(element) !== 0)
            return false; // not optional
        const upper = this.getUpperBound(element);
        return !upper.IsInfinity && upper.Value === 1;
    }
    isRequiredAndSinglevalued(element) {
        if (this.getLowerBound(element) !== 1)
            return false; // not required
        const upper = this.getUpperBound(element);
        return !upper.IsInfinity && upper.Value === 1;
    }
    // ********************  Default values **************************** //     
    getDefault(hasDefaultValue) {
        if (!hasDefaultValue.defaultValue)
            return null;
        return this.getValue(hasDefaultValue.defaultValue);
    }
    // ********************  ValueSpecification **************************** //     
    getValue(valueSpecification) {
        if ("value" in valueSpecification) { // don't use hasOwnProperty here
            // The valueSpecification is a LiteralInteger, LiteralBoolean, etc.
            return valueSpecification["value"];
        }
        else
            return null; // The valueSpecification is a LiteralNull or an unsupported implementation
    }
    getStringValue(valueSpecification) {
        const rawValue = valueSpecification.getValue();
        if (rawValue === null)
            return null;
        if (valueSpecification.elementType === Interfaces.ElementType.literalUnlimitedNatural) {
            return rawValue.stringValue();
        }
        return rawValue.toString();
    }
    // ********************  EnumerationLiteral **************************** //     
    getEnumeration(literal) {
        return literal.owner;
    }
    getSpecificationValue(literal) {
        if (!literal.specification)
            return null;
        return this.getValue(literal.specification);
    }
    // ********************  Property **************************** //     
    getAssociation(property) {
        // The property can be owned by an association (association.ownedEnds) or by a classifier
        // If owned by an association, the path is short.
        if (utils.isAssociation(property.owner)) {
            return property.owner;
        }
        // The property is not owned by an association, but can still be part of one
        // (association.memberEnds)
        if (!this.elementMap)
            return null;
        return this.elementMap.getAssociationHavingMemberEnd(property);
    }
    // ********************  Operation **************************** //     
    getReturnParameter(operation) {
        const val = operation.ownedParameters.find(p => { return p.direction === Interfaces.ParameterDirectionKind.return; });
        return val || null;
    }
    getInputParameters(operation) {
        return operation.ownedParameters.filter(p => {
            return p.direction === Interfaces.ParameterDirectionKind.in
                || p.direction === Interfaces.ParameterDirectionKind.inout;
        });
    }
    getOutputParameters(operation) {
        return operation.ownedParameters.filter(p => {
            return p.direction === Interfaces.ParameterDirectionKind.out
                || p.direction === Interfaces.ParameterDirectionKind.inout
                || p.direction === Interfaces.ParameterDirectionKind.return;
        });
    }
    getReturnType(operation) {
        const returnParameter = this.getReturnParameter(operation);
        return returnParameter ? returnParameter.type : null;
    }
}
