// @yellicode/elements/dist/es6/profile-utility.js
/*
 * Copyright (c) 2019 Yellicode
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import * as Interfaces from "./interfaces";
export class ProfileUtility {
    static getStereotypes(profile) {
        if (!profile.packagedElements)
            return [];
        return profile.packagedElements.filter((e) => e.elementType === Interfaces.ElementType.stereotype);
    }
    static hasStereotypeId(element, stereotypeId) {
        if (!element || !element.appliedStereotypes)
            return false;
        return ProfileUtility.hasStereotypeIdRecursive(element.appliedStereotypes, stereotypeId);
    }
    static hasStereotypeIdRecursive(stereotypes, stereotypeId) {
        for (var index = 0; index < stereotypes.length; index++) {
            var stereotype = stereotypes[index];
            if (stereotype.id === stereotypeId)
                return true;
            // If the stereotype has parents, check these too                        
            const stereotypeGenerals = stereotype.getParents();
            if (stereotypeGenerals.length > 0 && ProfileUtility.hasStereotypeIdRecursive(stereotypeGenerals, stereotypeId)) {
                return true;
            }
        }
        return false;
    }
    static getMetaClassesExtendedBy(stereotype) {
        // First add the stereotype's own meta classes
        return stereotype.extends.map(ext => ext.metaClass);
    }
    static getAllMetaClassesExtendedBy(stereotype) {
        // First add the stereotype's own meta classes
        const metaClasses = ProfileUtility.getMetaClassesExtendedBy(stereotype);
        // Then add meta classes of the specializing stereotypes
        stereotype.getSpecializations().forEach((derivedStereotype) => {
            derivedStereotype.extends.forEach(extension => {
                if (metaClasses.indexOf(extension.metaClass) === -1) {
                    metaClasses.push(extension.metaClass);
                }
            });
        });
        return metaClasses;
    }
    static hasProfileId(pack, profileId) {
        if (!pack || !pack.appliedProfiles)
            return false;
        return pack.appliedProfiles.some((s) => s.id === profileId);
    }
    /**
     * Filters the array of elements by only including the elements that have a particular stereotype applied.
     */
    static filterByStereotypeId(elements, stereotypeId, elementType) {
        if (!elements)
            return [];
        if (elementType) {
            return elements.filter(e => {
                return e.elementType === elementType && ProfileUtility.hasStereotypeId(e, stereotypeId);
            });
        }
        return elements.filter(e => { return ProfileUtility.hasStereotypeId(e, stereotypeId); });
    }
    /**
     * Filters the array of packageable element by only including the packages that have a particular profile applied.
     */
    static filterByProfileId(elements, profileId) {
        if (!elements)
            return [];
        return elements.filter((e) => e.elementType === Interfaces.ElementType.package &&
            ProfileUtility.hasProfileId(e, profileId));
    }
}
