// @yellicode/elements/dist/es6/profile-extender.js
/*
 * Copyright (c) 2019 Yellicode
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import { StringUtility } from '@yellicode/core';
import { ElementTypeUtility } from './utils';
import * as Interfaces from "./interfaces";
import * as Classes from "./classes";
import { ProfileUtility } from './profile-utility';
export class ProfileExtender {
    static applyProfiles(profiles) {
        return; // disabled in favor of modeler-generated profile API code
        // profiles.forEach(p => {
        //     ProfileExtender.applyProfile(p as Interfaces.Profile);
        // })        
    }
    static applyProfile(profile) {
        return; // disabled in favor of modeler-generated profile API code
        // const stereotypes = ProfileUtility.getStereotypes(profile);
        // if (stereotypes.length === 0) 
        //     return;
        // ProfileExtender.addHasProfileGetters(profile);
        // ProfileExtender.addHasStereotypeGetters(stereotypes);
    }
    /**
     * Extends all Package class types with getters for checking if the profile is applied to the Package.
     */
    static addHasProfileGetters(profile) {
        const classTypes = ProfileExtender.getClassTypesByElementType(Interfaces.ElementType.package);
        ProfileExtender.extendTypesWithGetter(classTypes, `hasProfile_${profile.safeName || profile.name}`, function () {
            return ProfileUtility.hasProfileId(this, profile.id);
        });
    }
    /**
     * Extends all extendable class types with getters for checking if a stereotype is applied.
     */
    static addHasStereotypeGetters(stereoTypes) {
        stereoTypes.forEach(st => {
            const metaClasses = ProfileUtility.getMetaClassesExtendedBy(st);
            metaClasses.forEach(metaClass => {
                const classTypes = ProfileExtender.getClassTypesByElementType(metaClass);
                ProfileExtender.extendTypesWithGetter(classTypes, `is_${st.safeName || st.name}`, function () {
                    // console.log(`Checking stereotype ${st.name}`);                   
                    return ProfileUtility.hasStereotypeId(this, st.id);
                });
                ProfileExtender.extendTypesWithGetter(classTypes, `as_${st.safeName || st.name}`, function () {
                    // console.log(`Checking stereotype ${st.name}`);                   
                    return this; // just return the current instance, this getter is just syntactic sugar for the TS compiler
                });
            });
        });
    }
    static addSubElementAccessors(stereoTypes, elementType, elementSubTypes) {
        const classTypes = ProfileExtender.getClassTypesByElementType(elementType);
        ProfileExtender.addSubSelementAccessorsForElementTypes(classTypes, stereoTypes, elementSubTypes);
    }
    /**
     * Adds accessor functions for stereotypes to the package (and model) level, e.g. getMyStereotypeClasses(),
     * getMyStereotypeInterfaces(), etc
     */
    static addPackageStereotypeAccessors(profile, stereoTypes) {
        // We apply the accessors to package types
        const packageTypes = ProfileExtender.getClassTypesByElementType(Interfaces.ElementType.package);
        stereoTypes.forEach(st => {
            // Does the stereotype extend a packageable element? If not, don't add an accessor on the package level.            
            const subPackagedElementTypes = st.extends.filter((ext) => ElementTypeUtility.isPackageableElement(ext.metaClass)).map(ext => ext.metaClass);
            if (subPackagedElementTypes.length === 0)
                return;
            // Add an accessor for each packageable element that the profile extends
            ProfileExtender.addSubSelementAccessorsForElementTypes(packageTypes, stereoTypes, subPackagedElementTypes);
            // Create a generic accessor for all packaged elements having the stereotype applied            
            ProfileExtender.extendTypesWithFunction(packageTypes, `getPackagedElementsOf${st.safeName || st.name}`, function () {
                return ProfileUtility.filterByStereotypeId(this.packagedElements, st.id);
            });
        });
        // Create a accessor for access to packages having this profile applied (e.g. getPackagesOfMyProfile)        
        ProfileExtender.extendTypesWithFunction(packageTypes, `getPackagesOf${profile.safeName || profile.name}`, function () {
            return ProfileUtility.filterByProfileId(this.packagedElements, profile.id);
        });
    }
    /**
     * Extends all classifier types with accessors to profile-specific stereotypes for Operations, Properties etc.
     */
    static addClassifierAccessors(stereoTypes) {
        //  All these types are a MemberedClassifier, so can have Properties (OwnedAttributes) and OwnedOperations    
        const classifierSubTypes = [Interfaces.ElementType.property, Interfaces.ElementType.operation];
        stereoTypes.forEach(st => {
            const memberedClassifiersExtendedByStereotype = st.extends.filter((ext) => ElementTypeUtility.isMemberedClassifier(ext.metaClass));
            memberedClassifiersExtendedByStereotype.forEach((extension) => {
                //  console.info(`Stereotype ${st.name} extends classifier ${StringUtil.capitalize(Interfaces.ElementType[extendedElementType])}`);
                const classTypes = ProfileExtender.getClassTypesByElementType(extension.metaClass);
                ProfileExtender.addSubSelementAccessorsForElementTypes(classTypes, stereoTypes, classifierSubTypes);
            });
        });
    }
    static addSubSelementAccessorsForElementTypes(classTypes, stereoTypes, elementTypes) {
        const classTypesDebugString = classTypes.map(t => t.name).join(', ');
        elementTypes.forEach(subElementType => {
            const stereoTypesExtendingSubType = stereoTypes.filter(s => s.extends.findIndex(t => t.metaClass === subElementType) > -1);
            if (stereoTypesExtendingSubType.length > 0) {
                // console.info(`ProfileExtender: Adding ${StringUtil.capitalize(Interfaces.ElementType[subElementType])} stereotype accessors to ${classTypesDebugString}.`);
            }
            else {
                // There are no Stereotypes that extend the subType, so there is no need to add an accessor
                // console.info(`ProfileExtender: Not adding ${StringUtil.capitalize(Interfaces.ElementType[subElementType])} stereotype accessors to ${classTypesDebugString}. There is no stereotype extending this element type.`);
                return;
            }
            // If this an "Owned" sub element (e.g. OwnedOperations, OwnedAttributes), add a prefix.
            const prefix = ElementTypeUtility.isPackageableElement(subElementType) ? '' : 'Owned';
            stereoTypesExtendingSubType.forEach(st => {
                const pluralElementTypeName = ProfileExtender.getGetPluralElementTypeName(subElementType);
                const functionName = `get${prefix}${pluralElementTypeName}Of${st.safeName || st.name}`; // E.g "getClassesOfMyStereotype"                        
                const accessorFunction = ProfileExtender.getAccessorFunctionForSubElementType(subElementType, st);
                if (accessorFunction) {
                    ProfileExtender.extendTypesWithFunction(classTypes, functionName, accessorFunction);
                }
                else {
                    // console.warn(`Failed adding accessor function '${functionName}' to ${classTypesDebugString}. Sub element type ${Interfaces.ElementType[subElementType]} is not supported.`);                
                }
            });
        });
    }
    static getAccessorFunctionForSubElementType(subElementType, stereoType) {
        if (ElementTypeUtility.isPackageableElement(subElementType)) {
            // E.g. getMyStereotypeClasses
            return function () {
                return ProfileUtility.filterByStereotypeId(this.packagedElements, stereoType.id, subElementType);
            };
        }
        switch (subElementType) {
            case Interfaces.ElementType.property:
                return function () {
                    return ProfileUtility.filterByStereotypeId(this.ownedAttributes, stereoType.id);
                };
            case Interfaces.ElementType.operation:
                return function () {
                    return ProfileUtility.filterByStereotypeId(this.ownedOperations, stereoType.id);
                };
            case Interfaces.ElementType.parameter:
                return function () {
                    return ProfileUtility.filterByStereotypeId(this.ownedParameters, stereoType.id);
                };
            case Interfaces.ElementType.enumerationLiteral:
                return function () {
                    return ProfileUtility.filterByStereotypeId(this.ownedLiterals, stereoType.id);
                };
            default:
                return null;
        }
    }
    static extendTypesWithFunction(classTypes, funcName, func) {
        classTypes.forEach(t => {
            // console.info(`ProfileExtender: Extending type ${t.name} with function '${funcName}'.`);           
            t.prototype[funcName] = func;
        });
    }
    static extendTypesWithGetter(classTypes, propertyName, getter) {
        classTypes.forEach(t => {
            // console.info(`ProfileExtender: Extending type ${t.name} with getter '${propertyName}'.`);
            Object.defineProperty(t.prototype, propertyName, {
                get: getter,
                enumerable: true,
                configurable: true,
            });
        });
    }
    /**
     * Gets the actual Javascript implementation types that match the specified element type.
     */
    static getClassTypesByElementType(elementType) {
        switch (elementType) {
            case Interfaces.ElementType.package:
                return [Classes.Model, Classes.Package];
            case Interfaces.ElementType.class:
                return [Classes.Class];
            case Interfaces.ElementType.dataType:
                return [Classes.DataType, Classes.PrimitiveType];
            case Interfaces.ElementType.interface:
                return [Classes.Interface];
            case Interfaces.ElementType.enumeration:
                return [Classes.Enumeration];
            case Interfaces.ElementType.enumerationLiteral:
                return [Classes.EnumerationLiteral];
            case Interfaces.ElementType.property:
                return [Classes.Property];
            case Interfaces.ElementType.operation:
                return [Classes.Operation];
            case Interfaces.ElementType.parameter:
                return [Classes.Parameter];
            default:
                console.warn(`Failed getting class type(s) for element type '${Interfaces.ElementType[elementType]}'. The element type is not supported.`);
                return [];
        }
    }
    static getGetPluralElementTypeName(elementType) {
        switch (elementType) {
            case Interfaces.ElementType.property:
                return 'Attributes';
            case Interfaces.ElementType.class:
                return 'Classes';
            case Interfaces.ElementType.enumerationLiteral:
                return 'Literals';
            default:
                // By default, add ..s
                var singular = StringUtility.capitalize(Interfaces.ElementType[elementType]);
                return `${singular}s`;
        }
    }
}
