// @yellicode/core/dist/es5/logger.js
/*
 * Copyright (c) 2019 Yellicode
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
/**
 * Enumerates the possible Yellicode logging levels.
 */
export var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["None"] = 0] = "None";
    LogLevel[LogLevel["Error"] = 1] = "Error";
    LogLevel[LogLevel["Warning"] = 2] = "Warning";
    LogLevel[LogLevel["Info"] = 3] = "Info";
    LogLevel[LogLevel["Verbose"] = 4] = "Verbose";
})(LogLevel || (LogLevel = {}));
(function (LogLevel) {
    function parse(str) {
        if (!str)
            return null;
        switch (str.toLowerCase()) {
            case 'none': return LogLevel.None;
            case 'error': return LogLevel.Error;
            case 'warning': return LogLevel.Warning;
            case 'info': return LogLevel.Info;
            case 'verbose': return LogLevel.Verbose;
            default: return null;
        }
    }
    LogLevel.parse = parse;
})(LogLevel || (LogLevel = {}));
/**
 * A Logger implementation that logs to the default Console.
 */
var ConsoleLogger = /** @class */ (function () {
    function ConsoleLogger(console, level, includeTimestamp) {
        if (level === void 0) { level = LogLevel.Info; }
        if (includeTimestamp === void 0) { includeTimestamp = false; }
        this.console = console;
        this.level = level;
        this.includeTimestamp = includeTimestamp;
    }
    ConsoleLogger.prototype.verbose = function (message) {
        if (this.level < LogLevel.Verbose)
            return;
        this.write("Verbose", message);
    };
    ;
    ConsoleLogger.prototype.info = function (message) {
        if (this.level < LogLevel.Info)
            return;
        this.write("Info", message);
    };
    ;
    ConsoleLogger.prototype.warn = function (message) {
        if (this.level < LogLevel.Warning)
            return;
        this.write("Warning", message);
    };
    ;
    ConsoleLogger.prototype.error = function (message) {
        if (this.level < LogLevel.Error)
            return;
        this.write("Error", message);
    };
    ;
    ConsoleLogger.prototype.log = function (message, level) {
        switch (level) {
            case LogLevel.Verbose:
                this.verbose(message);
                break;
            case LogLevel.Info:
                this.info(message);
                break;
            case LogLevel.Warning:
                this.warn(message);
                break;
            case LogLevel.Error:
                this.error(message);
                break;
        }
    };
    ConsoleLogger.prototype.write = function (level, message) {
        if (!this.includeTimestamp) {
            this.console.log(level + ": " + message);
            return;
        }
        var d = new Date();
        var time = d.toLocaleTimeString();
        var ms = ConsoleLogger.padMilliseconds(d.getMilliseconds());
        this.console.log(time + ":" + ms + " - " + level + ": " + message);
    };
    ConsoleLogger.padMilliseconds = function (num, length) {
        if (length === void 0) { length = 3; }
        if (('' + num).length >= length)
            return num;
        var lead = '0' + new Array(length).join('0');
        return (lead + num).slice(-length);
    };
    return ConsoleLogger;
}());
export { ConsoleLogger };
