// @yellicode/templating/dist/es6/stream-writer.js
/*
 * Copyright (c) 2019 Yellicode
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
const fs = require('fs');
const path = require('path');
const os = require('os');
import { DefaultRegionMarkerFormatter } from './region-marker-formatter';
/**
 * Internal writer class that is passed to WriterBase implementations. We don't let people inherit from TextWriter directly because
 * we want to control the instantiation and lifetime of each StreamWriter instance.
 */
export class StreamWriter {
    /**
     * Constructor. Creates a new StreamWriter that writes to the provided WritableStream.
     * @param stream A Node.js WritableStream instance.
     * @param regionMarkerFormatter An optional RegionMarkerFormatter.
     */
    constructor(stream, regionMarkerFormatter) {
        this.stream = stream;
        this.indent = 0;
        this.textFileCache = {};
        this.endOfLineString = os.EOL;
        this.indentString = "\t";
        this.templateDirName = path.resolve('.');
        this.fileRegionMapper = regionMarkerFormatter == null ? new DefaultRegionMarkerFormatter() : regionMarkerFormatter;
    }
    /******************************************************************************
     *                          TextWriter implementation
    ******************************************************************************/
    /**
    * Writes a new line to the output. The line is indented automatically. The line is ended with the endOfLineString.
    * @param value The line to write. When omitted, only the endOfLineString is written.
    */
    writeLine(value) {
        if (value) {
            this.stream.write(this.createIndentString() + value + this.endOfLineString);
        }
        else
            this.stream.write(this.endOfLineString);
        return this;
    }
    ;
    /**
    * Writes a collection of lines to the output. Each line is indented automatically and ended with the endOfLineString.
    * @param values The lines to write.
    * @param delimiter An optional delimiter to be written at the end of each line, except for the last one.
    */
    writeLines(values, delimiter) {
        if (!values)
            return this;
        const len = values.length;
        values.forEach((value, index) => {
            this.stream.write(this.createIndentString() + value);
            if (delimiter && index < len - 1) {
                this.stream.write(delimiter);
            }
            this.stream.write(this.endOfLineString);
        });
        return this;
    }
    /**
     * Writes a new line to the output while temporarily increasing the indent. The line is ended with the endOfLineString.
     * @param value The line to write.
     */
    writeLineIndented(value) {
        this.stream.write(this.indentString + this.createIndentString() + value + this.endOfLineString);
        return this;
    }
    writeEndOfLine(value) {
        if (value) {
            this.stream.write(value);
        }
        this.stream.write(this.endOfLineString);
        return this;
    }
    writeIndent() {
        this.stream.write(this.createIndentString());
        return this;
    }
    ;
    write(value) {
        if (value == null)
            return this; // avoid 'May not write null values to stream'
        this.stream.write(value);
        return this;
    }
    ;
    writeWhiteSpace() {
        this.write(" ");
        return this;
    }
    increaseIndent() {
        this.indent++;
        return this;
    }
    decreaseIndent() {
        if (this.indent > 0) {
            this.indent--;
        }
        return this;
    }
    clearIndent() {
        this.indent = 0;
        return this;
    }
    writeFile(path, encoding) {
        const fullPath = this.resolveFileName(path);
        const contents = this.readTextFile(fullPath, false, encoding);
        if (contents == null || contents.length <= 0)
            return this;
        this.stream.write(contents);
        return this;
    }
    writeFileRegion(regionName, path, encoding) {
        if (this.fileRegionMapper == null || regionName == null || regionName.length === 0 || path == null || path.length === 0)
            return false;
        const fullPath = this.resolveFileName(path);
        const contents = this.readTextFile(fullPath, true, encoding);
        if (contents == null || contents.length <= 0)
            return false;
        // First find the region match 
        const regionStartMarker = this.fileRegionMapper.getRegionStartMarker(regionName);
        const regionStartIndex = contents.indexOf(regionStartMarker);
        if (regionStartIndex < 0) {
            return false;
        }
        const regionEndMarker = this.fileRegionMapper.getRegionEndMarker(regionName);
        const regionEndIndex = contents.indexOf(regionEndMarker, regionStartIndex);
        if (regionEndIndex < 0) {
            return false;
        }
        var region = contents.substring(regionStartIndex + regionStartMarker.length, regionEndIndex);
        this.stream.write(region);
        this.writeEndOfLine();
        return true;
    }
    /******************************************************************************
     *                         End of TextWriter implementation
    ******************************************************************************/
    createIndentString() {
        var result = "";
        for (let i = 0; i < this.indent; i++) {
            result += this.indentString;
        }
        return result;
    }
    ;
    resolveFileName(fileName) {
        return path.join(this.templateDirName, fileName);
    }
    readTextFile(path, useCache, encoding) {
        if (useCache && this.textFileCache.hasOwnProperty(path)) {
            return this.textFileCache[path];
        }
        if (!fs.existsSync(path)) {
            //  this.writeLine(`Cannot write the contents of file '${fullPath}' because the file does not exist.`);
            if (useCache) {
                this.textFileCache[path] = null;
            }
            return undefined;
        }
        if (encoding != null) {
            encoding = encoding.toLowerCase();
        }
        else
            encoding = 'utf-8';
        var contents = fs.readFileSync(path, encoding);
        if (contents == null || contents.length === 0)
            return undefined;
        // Remove any utf-8 BOM if there is any
        if (encoding === 'utf-8' && contents.charCodeAt(0) === 0xFEFF) {
            contents = contents.slice(1);
        }
        if (useCache) {
            this.textFileCache[path] = contents;
        }
        return contents;
    }
}
